R/ml_replicate.R

```diff
@@ -0,0 +1,144 @@
+#' Replicate records in a reference sample based on its fitted weights
+#'
+#' @description
+#' This function replicates each entry in a reference sample based on its fitted
+#' weights. This is useful if the result of multiple replication algorithms 
+#' are compared to each other, or to generate a full synthetic population 
+#' based on the result of a `ml_fit` object. Note that, all individual 
+#' and group ids of the synthetic population are not the same as those in
+#' the original reference sample, and the total number of groups replicated
+#' is always very close to or equal the sum of the fitted group weights.
+#'
+#' @param ml_fit A `ml_fit` object created by the [ml_fit()] family.
+#' @param algorithm Replication algorithm to use. "trs" is
+#'  the 'Truncate, replicate, sample' integerisation algorithm proposed
+#'  by Lovelace et al. (2013), "pp" is weighted sampling with
+#'  replacement, and "round" is just simple rounding.
+#' @param verbose If `TRUE`, print diagnostic output.
+#' @param .keep_original_ids If `TRUE`, the original individual and group
+#'  ids of the reference sample will be kept with suffix '_old'.
+#'
+#' @references
+#'      Lovelace, R., & Ballas, D. (2013). ‘Truncate, replicate, sample’:
+#'          A method for creating integer weights for spatial microsimulation.
+#'          Computers, Environment and Urban Systems, 41, 1-11.
+#'
+#' @return All functions return a data.frame.
+#'
+#' @export
+#' @examples
+#' path <- toy_example("Tiny")
+#' fit <- ml_fit(algorithm = "entropy_o", fitting_problem = readRDS(path))
+#' syn_pop <- ml_replicate(fit, algorithm = "trs")
+#' syn_pop
+ml_replicate <- function(ml_fit, algorithm = c("pp", "trs", "round"), verbose = FALSE, .keep_original_ids = FALSE) {
+  algorithm <- match.arg(algorithm)
+  fun.name <- sprintf("int_%s", algorithm)
+  if (!exists(fun.name)) {
+    stop("Unknown replication algorithm:", algorithm)
+  }
+
+  .patch_verbose()
+
+  message("Replicate using '", algorithm, "' algorithm")
+  group_id <- ml_fit$flat$fitting_problem$fieldNames$groupId
+  count_col <- ml_fit$flat$fitting_problem$fieldNames$count
+
+  message("Extracting fitted weights of each group")
+  number_of_persons_in_each_group <-
+    ml_fit$flat$fitting_problem$refSample[[group_id]] %>%
+    table() %>%
+    as.integer()
+  weights <-
+    ml_fit$flat_weights
+
+  message("Integerising the fitted weights")
+  integerised_weights <- get(fun.name)(weights = weights)
```

Comment:
How about extracting a function that combines line 36 with the `get()` call? Do we need to scope the `get()` call to the package environment, and to return functions only, to avoid confusion with other objects named `int_trs`, e.g. in `.GlobalEnv` ?

Reply:
Sounds good. Will fix this.

Reply:
Apparently, `get()` can't find a non-exported function but `getFromNamespace()` can.